external-crates/move/crates/move-analyzer/trace-adapter/src/runtime.ts

```diff
@@ -198,41 +178,136 @@ export class Runtime extends EventEmitter {
         while (true) {
             this.eventIndex++;
             if (this.eventIndex >= this.trace.events.length) {
-                return true;
+                throw new Error("Cannot find corresponding CloseFrame event for function: " +
+                    currentFrame.name);
             }
             currentEvent = this.trace.events[this.eventIndex];
-            if (currentEvent.type === 'CloseFrame' && currentEvent.id === openFrameID) {
+            if (currentEvent.type === 'CloseFrame' && currentEvent.id === currentFrame.id) {
                 break;
             }
         }
         this.frameStack.frames.pop();
         return this.step(false);
+    }
+    /**
+     * Handles "step back" adapter action.
+     * @throws Error with a descriptive error message if the step back event cannot be handled.
+     */
+    public stepBack() {
+        if (this.eventIndex === 1) {
+            // no where to step back to (event 0 is the `OpenFrame` event for the first frame)
+            // and is processed in runtime.start() which is executed only once
+            this.sendEvent(RuntimeEvents.stopOnStep);
+            return;
+        }
+        let currentEvent = this.trace.events[this.eventIndex - 1];
+        if (currentEvent.type === 'CloseFrame') {
+            // cannot step back into or over function calls
+            this.sendEvent(RuntimeEvents.stopOnStep);
+            return;
+        } else {
+            this.eventIndex--;
+            if (currentEvent.type === 'Instruction') {
+                let sameLine = this.instruction(currentEvent);
+                if (sameLine) {
+                    this.stepBack();
+                    return;
+                }
+                this.sendEvent(RuntimeEvents.stopOnStep);
+                return;
+            } else if (currentEvent.type === 'OpenFrame') {
+                // pop the top frame from the stack
+                this.frameStack.frames.pop();
+                // cannot simply call stepBack as we are stepping back to the same line
+                // that is now in the current frame, which would result in unintentionally
+                // recursing to previous events
+                if (this.eventIndex === 0) {
+                    // no where to step back to
+                    this.sendEvent(RuntimeEvents.stopOnStep);
+                    return;
+                }
+                this.eventIndex--;
+                let currentEvent = this.trace.events[this.eventIndex];
+                if (currentEvent.type !== 'Instruction') {
+                    throw new Error("Expected an Instruction event before OpenFrame event");
+                }
+                if (!this.instruction(currentEvent)) {
+                    // we should be steppping back to the instruction on the same line
+                    // as the one in the current frame
+                    throw new Error("Wrong line to step back to from a function call");
+                }
```

Comment:
Minor: would it be helpful to include the `eventIndex` in the error?

Reply:
Good idea! I re-worked error messages in this function more generally to make the more descriptive